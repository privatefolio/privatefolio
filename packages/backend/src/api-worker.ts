// import "./comlink-setup"

import { expose } from "comlink"

import { api } from "./api/api"
// import { api as rawApi } from "./api/api"

// const api = new Proxy(rawApi, {
//   get(api, method: string) {
//     const orig = (api as unknown)[method]
//     return async (...args: unknown[]) => {
//       try {
//         return await orig.apply(api, args)
//       } catch (err) {
//         console.error(`[Worker] Error in ${method} ${String(err)}`)
//         // return "Something went wrong"
//         throw err
//       }
//     }
//   },
// })

expose(api)
console.log("Bun worker initialized.")
