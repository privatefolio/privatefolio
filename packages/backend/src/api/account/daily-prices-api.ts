import { PAIR_MAPPER, PRICE_APIS, PRICE_MAPPER } from "src/extensions/prices/providers"
import {
  ChartData,
  DailyPrice,
  MyAsset,
  ProgressCallback,
  ResolutionString,
  SqlParam,
  SubscriptionChannel,
  TaskPriority,
  TaskTrigger,
  Time,
  Timestamp,
} from "src/interfaces"
import { allPriceApiIds } from "src/settings/price-apis"
import { PRICE_API_PAGINATION, PRICE_APIS_META } from "src/settings/settings"
import { getAssetTicker } from "src/utils/assets-utils"
import { formatDate } from "src/utils/formatting-utils"
import { createSubscription } from "src/utils/sub-utils"
import { isTestEnvironment, noop } from "src/utils/utils"

import { getAccount } from "../accounts-api"
import { getMyAssets, patchAsset } from "./assets-api"
import { enqueueTask } from "./server-tasks-api"

type NewDailyPrice = Omit<DailyPrice, "id">

function deriveDailyPriceId(record: NewDailyPrice) {
  return `${record.assetId}_${record.timestamp}`
}

export async function upsertDailyPrices(
  accountName: string,
  records: DailyPrice[] | NewDailyPrice[]
) {
  const account = await getAccount(accountName)

  try {
    await account.executeMany(
      `INSERT OR REPLACE INTO daily_prices (
        id, assetId, timestamp, price, pair, priceApiId
      ) VALUES (?, ?, ?, ?, ?, ?)`,
      records.map((record) => [
        "id" in record ? record.id : deriveDailyPriceId(record),
        record.assetId,
        record.timestamp,
        JSON.stringify(record.price) || null,
        record.pair || null,
        record.priceApiId || null,
      ])
    )
    account.eventEmitter.emit(SubscriptionChannel.DailyPrices)
  } catch (error) {
    throw new Error(`Failed to add or replace daily prices: ${error}`)
  }
}

export async function upsertDailyPrice(accountName: string, record: DailyPrice | NewDailyPrice) {
  return upsertDailyPrices(accountName, [record])
}

export async function getPricesForAsset(
  accountName: string,
  assetId: string,
  timestamp?: Timestamp,
  start?: Timestamp,
  end?: Timestamp
): Promise<ChartData[]> {
  if (assetId === "USDT" && !!timestamp) {
    return [{ time: timestamp / 1000, value: 1 }] as ChartData[]
  }

  const account = await getAccount(accountName)

  let query = `
    SELECT price FROM daily_prices
    WHERE assetId = ? 
  `
  if (timestamp && (start || end)) {
    throw new Error("You can only pass timestamp or start/end, not both")
  }

  const params: SqlParam[] = []

  if (timestamp) {
    query += " AND timestamp = ?"
    params.push(timestamp)
  }
  if (start && end) {
    query += " AND timestamp >= ? AND timestamp <= ?"
    params.push(start, end)
  } else if (start) {
    query += " AND timestamp >= ?"
    params.push(start)
  } else if (end) {
    query += " AND timestamp <= ?"
    params.push(end)
  }

  query += " ORDER BY timestamp ASC"

  const prices = await account.execute(query, [assetId, ...params])

  return prices.map((x) => JSON.parse(x[0] as string))
}

export async function getAssetPriceMap(
  accountName: string,
  timestamp: Timestamp = new Date().getTime()
): Promise<Record<string, ChartData>> {
  const account = await getAccount(accountName)

  const day: Timestamp = timestamp - (timestamp % 86400000)

  const prices = await account.execute(
    `
    SELECT assetId, price FROM daily_prices
    WHERE timestamp = ?
  `,
    [day]
  )

  return prices.reduce(
    (map: Record<string, ChartData>, x) => {
      map[x[0] as string] = JSON.parse(x[1] as string)
      return map
    },
    {
      USDT: { time: (day / 1000) as Time, value: 1 },
    }
  )
}

export async function getPriceCursor(
  accountName: string,
  assetId: string
): Promise<Timestamp | undefined> {
  const account = await getAccount(accountName)

  const query = `
    SELECT timestamp FROM daily_prices
    WHERE assetId = ?
    ORDER BY timestamp DESC
    LIMIT 1
  `
  const prices = await account.execute(query, [assetId])

  if (prices.length === 0) {
    return undefined
  }

  return JSON.parse(prices[0][0] as string)
}

type FetchDailyPricesOptions = {
  /**
   * less than or equal to this timestamp
   */
  until?: number
}

export async function fetchDailyPrices(
  accountName: string,
  assetsParam?: MyAsset[],
  progress: ProgressCallback = noop,
  signal?: AbortSignal,
  options: FetchDailyPricesOptions = {}
) {
  let assets: MyAsset[]

  if (!assetsParam) {
    assets = await getMyAssets(accountName)
  } else {
    assets = assetsParam
  }

  await progress([0, `Fetching asset prices for ${assets.length} assets`])

  const now = Date.now()
  const today: Timestamp = now - (now % 86400000)

  const promises: (() => Promise<void>)[] = []

  for (let i = 1; i <= assets.length; i++) {
    // eslint-disable-next-line no-loop-func
    promises.push(async () => {
      const asset = assets[i - 1]

      if (!asset.coingeckoId && !isTestEnvironment) {
        await progress([undefined, `Skipped ${getAssetTicker(asset.id)}: No coingeckoId`])
        return
      }

      if (signal?.aborted) throw new Error(signal.reason)

      const preferredPriceApiId = asset.priceApiId
      const priceApiIds = preferredPriceApiId ? [preferredPriceApiId] : allPriceApiIds

      let since: Timestamp | undefined = await getPriceCursor(accountName, asset.id)
      let until: Timestamp | undefined = options.until || today

      if (!since) since = until - 86400000 * PRICE_API_PAGINATION - 1
      if (since === until) since = since - 1

      for (const priceApiId of priceApiIds) {
        try {
          while (true) {
            const priceApi = PRICE_APIS[priceApiId]
            const priceMapper = PRICE_MAPPER[priceApiId]
            const pairMapper = PAIR_MAPPER[priceApiId]

            if (!priceApi || !priceMapper || !pairMapper) {
              throw new Error(`Price API "${priceApiId}" is not supported`)
            }

            const pair = pairMapper(asset.id)

            const request = {
              limit: PRICE_API_PAGINATION,
              pair,
              since,
              timeInterval: "1d" as ResolutionString,
              until,
            }

            // console.log("Daily price api request:", request)

            const results = await priceApi(request)

            if (!preferredPriceApiId && results.length > 0) {
              patchAsset(accountName, asset.id, { priceApiId })
            }

            if (signal?.aborted) throw new Error(signal.reason)

            if (results.length === 0) {
              throw new Error(`${PRICE_APIS_META[priceApiId].name}: EmptyResponse`)
            }

            const documents = results.map((result) => {
              const price = priceMapper(result)
              const timestamp = (price.time as number) * 1000

              const doc: NewDailyPrice = {
                assetId: asset.id,
                pair,
                price,
                priceApiId,
                timestamp,
              }

              return doc
            })

            await upsertDailyPrices(accountName, documents)

            const start = (priceMapper(results[0]).time as number) * 1000
            const end = (priceMapper(results[results.length - 1]).time as number) * 1000

            await progress([
              undefined,
              `Fetched ${getAssetTicker(asset.id)} using ${
                PRICE_APIS_META[priceApiId].name
              } from ${formatDate(start)} to ${formatDate(end)}`,
            ])

            if (results.length !== PRICE_API_PAGINATION) {
              // reached listing date (genesis)
              break
            }

            until = start - 86400000
            since = start - 86400000 * PRICE_API_PAGINATION
          }
          break
        } catch (error) {
          await progress([undefined, `Skipped ${getAssetTicker(asset.id)}: ${error}`])
        }
      }
    })
  }

  let progressCount = 0

  await Promise.all(
    promises.map((fetchFn) =>
      fetchFn().then(async () => {
        progressCount += 1
        await progress([(progressCount / assets.length) * 100])
      })
    )
  )
}

export function enqueueFetchPrices(accountName: string, trigger: TaskTrigger) {
  return enqueueTask(accountName, {
    description: "Fetching price data for all assets.",
    determinate: true,
    function: async (progress, signal) => {
      await fetchDailyPrices(accountName, undefined, progress, signal)
    },
    name: "Fetch asset prices",
    priority: TaskPriority.Low,
    trigger,
  })
}

export async function deleteDailyPrices(accountName: string) {
  const account = await getAccount(accountName)
  await account.execute("DELETE FROM daily_prices")
}

export function enqueueDeleteAssetPrices(accountName: string, trigger: TaskTrigger) {
  return enqueueTask(accountName, {
    description: "Deleting price data for all assets.",
    determinate: true,
    function: async () => {
      await deleteDailyPrices(accountName)
    },
    name: "Delete asset prices",
    priority: TaskPriority.High,
    trigger,
  })
}

export async function subscribeToDailyPrices(accountName: string, callback: () => void) {
  return createSubscription(accountName, SubscriptionChannel.DailyPrices, callback)
}
