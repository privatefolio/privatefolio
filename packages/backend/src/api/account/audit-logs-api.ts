import { EventCause, SqlParam } from "src/interfaces"
import { transformNullsToUndefined } from "src/utils/db-utils"
import { createSubscription } from "src/utils/sub-utils"

import { AuditLog, AuditLogOperation, SubscriptionChannel } from "../../interfaces"
import { getAccount } from "../accounts-api"

export async function getAuditLogs(
  accountName: string,
  query = "SELECT * FROM audit_logs ORDER BY timestamp DESC, changeN DESC, id DESC",
  params?: SqlParam[]
): Promise<AuditLog[]> {
  const account = await getAccount(accountName)

  try {
    const result = await account.execute(query, params)
    return result.map((row) => {
      /* eslint-disable sort-keys-fix/sort-keys-fix */
      const value = {
        id: row[0],
        assetId: row[1],
        balance: row[2],
        change: row[4],
        fileImportId: row[6],
        connectionId: row[7],
        importIndex: row[8],
        operation: row[9],
        platform: row[10],
        timestamp: row[11],
        txId: row[12],
        wallet: row[13],
      }
      /* eslint-enable */
      transformNullsToUndefined(value)
      return value as AuditLog
    })
  } catch (error) {
    throw new Error(`Failed to query audit logs: ${error}`)
  }
}

export async function getAuditLog(accountName: string, id: string) {
  const records = await getAuditLogs(accountName, "SELECT * FROM audit_logs WHERE id = ?", [id])
  return records[0]
}

export async function getAuditLogsByTxId(accountName: string, txId: string) {
  return getAuditLogs(accountName, "SELECT * FROM audit_logs WHERE txId = ?", [txId])
}

export async function upsertAuditLogs(accountName: string, records: AuditLog[]) {
  const account = await getAccount(accountName)

  try {
    await account.executeMany(
      `INSERT OR REPLACE INTO audit_logs (
      id, assetId, balance, change, fileImportId, connectionId, 
      importIndex, operation, platform, timestamp, txId, wallet
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      records.map((record) => [
        record.id,
        record.assetId,
        record.balance || null,
        record.change,
        record.fileImportId || null,
        record.connectionId || null,
        record.importIndex,
        record.operation,
        record.platform,
        record.timestamp,
        record.txId || null,
        record.wallet,
      ])
    )
    // TODO8 this will help invalidating balances cursor
    // let earliestTimestamp = Date.now()
    // for (const record of records) {
    //   if (record.timestamp < earliestTimestamp) {
    //     earliestTimestamp = record.timestamp
    //   }
    // }

    account.eventEmitter.emit(SubscriptionChannel.AuditLogs, EventCause.Created)
  } catch (error) {
    throw new Error(`Failed to add or replace audit logs: ${error}`)
  }
}

export async function upsertAuditLog(accountName: string, record: AuditLog) {
  return upsertAuditLogs(accountName, [record])
}

export async function patchAuditLog(accountName: string, id: string, patch: Partial<AuditLog>) {
  const existing = await getAuditLog(accountName, id)
  const newValue = { ...existing, ...patch }
  await upsertAuditLog(accountName, newValue)
}

export async function countAuditLogs(
  accountName: string,
  query = "SELECT COUNT(*) FROM audit_logs",
  params?: SqlParam[]
): Promise<number> {
  const account = await getAccount(accountName)

  try {
    const result = await account.execute(query, params)
    return result[0][0] as number
  } catch (error) {
    throw new Error(`Failed to count audit logs: ${error}`)
  }
}

export async function subscribeToAuditLogs(
  accountName: string,
  callback: (cause: EventCause) => void
) {
  return createSubscription(accountName, SubscriptionChannel.AuditLogs, callback)
}

export async function getMyPlatformIds(
  accountName: string,
  query = "SELECT DISTINCT platform FROM audit_logs ORDER BY platform ASC",
  params?: SqlParam[]
): Promise<string[]> {
  const account = await getAccount(accountName)

  try {
    const result = await account.execute(query, params)
    return result.map((row) => row[0] as string)
  } catch (error) {
    throw new Error(`Failed to query platforms: ${error}`)
  }
}

export async function getWallets(
  accountName: string,
  query = "SELECT DISTINCT wallet FROM audit_logs ORDER BY wallet ASC",
  params?: SqlParam[]
): Promise<string[]> {
  const account = await getAccount(accountName)

  try {
    const result = await account.execute(query, params)
    return result.map((row) => row[0] as string)
  } catch (error) {
    throw new Error(`Failed to query wallets: ${error}`)
  }
}

export async function getOperations(
  accountName: string,
  query = "SELECT DISTINCT operation FROM audit_logs ORDER BY operation ASC",
  params?: SqlParam[]
): Promise<AuditLogOperation[]> {
  const account = await getAccount(accountName)

  try {
    const result = await account.execute(query, params)
    return result.map((row) => row[0] as AuditLogOperation)
  } catch (error) {
    throw new Error(`Failed to query operations: ${error}`)
  }
}
