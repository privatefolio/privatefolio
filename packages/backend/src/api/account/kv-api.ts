import { SqlParam, SubscriptionChannel } from "src/interfaces"

import { getAccount } from "../accounts-api"

export async function setValue(key: string, value: unknown, accountName: string) {
  const account = await getAccount(accountName)
  await account.execute("INSERT OR REPLACE INTO key_value (key, value) VALUES (?, ?)", [
    key,
    value as SqlParam,
  ])
  account.eventEmitter.emit(SubscriptionChannel.KeyValue, key, value)
}

export async function getValue<T>(
  accountName: string,
  key: string,
  defaultValue: T | null = null
): Promise<T | null> {
  const account = await getAccount(accountName)
  const result = await account.execute("SELECT * FROM key_value WHERE key = ?", [key])
  const existing = result[0] as unknown as [string, T][]

  return result.length > 0 ? (existing[1] as T) : defaultValue
}

export async function subscribeToKV<T>(
  accountName: string,
  key: string,
  callback: (value: T) => void
) {
  const account = await getAccount(accountName)

  function listener(k: string, value: T) {
    if (key === k) {
      callback(value)
    }
  }

  account.eventEmitter.on(SubscriptionChannel.KeyValue, listener)
  return () => account.eventEmitter.off(SubscriptionChannel.KeyValue, listener)
}
