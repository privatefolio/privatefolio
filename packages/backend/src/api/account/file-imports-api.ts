import { access, readFile } from "fs/promises"
import { join } from "path"
import { getAccount } from "src/api/accounts-api"
import { HEADER_MATCHER, PARSER_MATCHER } from "src/extensions/file-imports/matchers"
import { parseCsv, sanitizeHeader } from "src/extensions/file-imports/parse-csv"
import {
  EventCause,
  FileImport,
  ProgressCallback,
  ServerFile,
  SqlParam,
  SubscriptionChannel,
  TaskCompletionCallback,
  TaskPriority,
  TaskTrigger,
  Timestamp,
} from "src/interfaces"
import { FILES_LOCATION } from "src/settings/settings"
import { splitRows } from "src/utils/csv-utils"
import { formatDate } from "src/utils/formatting-utils"
import { createSubscription } from "src/utils/sub-utils"
import { hashString, noop } from "src/utils/utils"

import { countAuditLogs, upsertAuditLogs } from "./audit-logs-api"
import { invalidateBalances } from "./balances-api"
import { invalidateNetworth } from "./networth-api"
import { getServerFile } from "./server-files-api"
import { enqueueTask } from "./server-tasks-api"
import { invalidateTradePnl, invalidateTrades } from "./trades-api"
import { countTransactions, upsertTransactions } from "./transactions-api"

async function readFileImportText(accountName: string, fileRecord: ServerFile) {
  const filePath = join(FILES_LOCATION, accountName, String(fileRecord.id))
  await access(filePath)
  const text = await readFile(filePath, "utf-8")
  return text
}

export async function getFileImportRequirements(accountName: string, fileRecord: ServerFile) {
  const text = await readFileImportText(accountName, fileRecord)
  const rows = splitRows(text)
  const header = sanitizeHeader(rows[0])
  const parserId = HEADER_MATCHER[header]
  const parser = PARSER_MATCHER[parserId]
  return parser?.requirements
}

export async function importFile(
  accountName: string,
  fileRecord: ServerFile,
  parserContext: Record<string, unknown> = {},
  progress: ProgressCallback = noop
): Promise<FileImport> {
  const account = await getAccount(accountName)

  const { name } = fileRecord
  const { type, lastModified, size } = fileRecord.metadata

  await progress([0, `Reading file "${name}" from disk`])
  const text = await readFileImportText(accountName, fileRecord)
  await progress([10, `Read file "${name}" from disk`])

  if (type !== "text/csv") {
    throw new Error("Error reading file: not .csv")
  }

  // TESTME TODO2 Looks like on mobile, lastModified is set to Date.now()
  const id = hashString(`fi_${name}_${size}_${lastModified}`)
  // const id = hashString(`fi_${name}_${size}`)

  await progress([11, `Saving file import metadata under the id ${id}`])

  try {
    await account.execute(
      `INSERT INTO file_imports (id, lastModified, name, size) VALUES (?, ?, ?, ?)`,
      [id, lastModified, name, size]
    )
  } catch {
    throw new Error(`You cannot import the same file twice.`)
  }

  try {
    await progress([12, `Parsing file ${name}`])
    const { metadata, logs, transactions } = await parseCsv(text, id, progress, parserContext)

    await progress([60, `Saving ${logs.length} audit logs to disk`])
    await upsertAuditLogs(accountName, logs)

    await progress([80, `Saving ${transactions.length} transactions to disk`])
    await upsertTransactions(accountName, transactions)

    let oldestTimestamp: Timestamp | undefined
    for (const log of logs) {
      if (!oldestTimestamp || log.timestamp < oldestTimestamp) {
        oldestTimestamp = log.timestamp
      }
    }

    if (oldestTimestamp) {
      const newCursor = oldestTimestamp - (oldestTimestamp % 86400000) - 86400000
      await progress([25, `Setting balances cursor to ${formatDate(newCursor)}`])
      await invalidateBalances(accountName, newCursor)
      await invalidateNetworth(accountName, newCursor)
      await invalidateTrades(accountName, newCursor)
      await invalidateTradePnl(accountName, newCursor)
    }

    await account.execute(`UPDATE file_imports SET timestamp = ?, meta = ? WHERE id = ?`, [
      new Date().getTime(),
      JSON.stringify(metadata),
      id,
    ])

    const fileImport = await getFileImport(accountName, id)
    account.eventEmitter.emit(SubscriptionChannel.FileImports, EventCause.Created, fileImport)

    return fileImport
  } catch (error) {
    await account.execute(`DELETE FROM file_imports WHERE id = ?`, [id])
    throw error
  }
}

export async function getFileImports(
  accountName: string,
  query = "SELECT * FROM file_imports ORDER BY timestamp DESC",
  params?: SqlParam[]
) {
  const account = await getAccount(accountName)
  try {
    const result = await account.execute(query, params)
    return result.map((row) => ({
      id: row[0] as string,
      lastModified: row[1] as number,
      meta: JSON.parse(row[5] as string),
      name: row[2] as string,
      size: row[3] as number,
      timestamp: row[4] as number,
    })) satisfies FileImport[]
  } catch (error) {
    throw new Error(`Failed to query file imports: ${error}`)
  }
}
export async function getFileImport(accountName: string, id: string) {
  const records = await getFileImports(accountName, "SELECT * FROM file_imports WHERE id = ?", [id])
  return records[0]
}

export async function countFileImports(
  accountName: string,
  query = "SELECT COUNT(*) FROM file_imports",
  params?: SqlParam[]
): Promise<number> {
  const account = await getAccount(accountName)

  try {
    const result = await account.execute(query, params)
    return result[0][0] as number
  } catch (error) {
    throw new Error(`Failed to count file imports: ${error}`)
  }
}

export async function deleteFileImport(
  accountName: string,
  fileImport: FileImport,
  progress: ProgressCallback = noop
) {
  await progress([0, `Removing file import with id ${fileImport.id}`])
  const account = await getAccount(accountName)

  const oldestTimestamp = (
    await account.execute(
      `SELECT timestamp FROM audit_logs WHERE fileImportId = ? ORDER BY timestamp ASC LIMIT 1`,
      [fileImport.id]
    )
  )[0]?.[0] as Timestamp | undefined

  const auditLogsCount = await countAuditLogs(
    accountName,
    `SELECT COUNT(*) FROM audit_logs WHERE fileImportId = ?`,
    [fileImport.id]
  )
  await progress([0, `Removing ${auditLogsCount} audit logs`])
  await account.execute(`DELETE FROM audit_logs WHERE fileImportId = ?`, [fileImport.id])
  account.eventEmitter.emit(SubscriptionChannel.AuditLogs, EventCause.Deleted)

  if (oldestTimestamp) {
    const newCursor = oldestTimestamp - (oldestTimestamp % 86400000) - 86400000
    await progress([25, `Setting balances cursor to ${formatDate(newCursor)}`])
    await invalidateBalances(accountName, newCursor)
    await progress([25, `Setting networth cursor to ${formatDate(newCursor)}`])
    await invalidateNetworth(accountName, newCursor)
  }

  const transactionsCount = await countTransactions(
    accountName,
    `SELECT COUNT(*) FROM transactions WHERE fileImportId = ?`,
    [fileImport.id]
  )
  await progress([50, `Removing ${transactionsCount} transactions`])
  await account.execute(`DELETE FROM transactions WHERE fileImportId = ?`, [fileImport.id])

  await account.execute(`DELETE FROM file_imports WHERE id = ?`, [fileImport.id])
  await progress([100, `Removed file import with id ${fileImport.id}`])

  account.eventEmitter.emit(SubscriptionChannel.FileImports, EventCause.Deleted, fileImport)

  return auditLogsCount
}

export async function subscribeToFileImports(
  accountName: string,
  callback: (cause: EventCause, fileImport: FileImport) => void
) {
  return createSubscription(accountName, SubscriptionChannel.FileImports, callback)
}

export async function enqueueImportFile(
  accountName: string,
  trigger: TaskTrigger,
  fileId: number,
  parserContext?: Record<string, unknown>
) {
  const fileRecord = await getServerFile(accountName, fileId)
  return enqueueTask(accountName, {
    description: `Importing "${fileRecord.name}".`,
    determinate: true,
    function: (progress) => importFile(accountName, fileRecord, parserContext, progress),
    name: `Import "${fileRecord.name}"`,
    priority: TaskPriority.VeryHigh,
    trigger,
  })
}

export async function enqueueRemoveFileImport(
  accountName: string,
  trigger: TaskTrigger,
  fileImportId: string,
  onCompletion?: TaskCompletionCallback
) {
  const fileImport = await getFileImport(accountName, fileImportId)
  return enqueueTask(accountName, {
    description: `Remove "${fileImport.name}", alongside its audit logs and transactions.`,
    determinate: true,
    function: async (progress) => {
      await deleteFileImport(accountName, fileImport, progress)
    },
    name: `Remove file import`,
    onCompletion,
    priority: TaskPriority.VeryHigh,
    trigger,
  })
}
