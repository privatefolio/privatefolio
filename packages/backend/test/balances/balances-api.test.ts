import fs from "fs"
import { join } from "path"
import { getAuditLogs } from "src/api/account/audit-logs-api"
import {
  computeBalances,
  deleteBalances,
  getBalances,
  getBalancesAt,
} from "src/api/account/balances-api"
import { fetchDailyPrices } from "src/api/account/daily-prices-api"
import { importFile } from "src/api/account/file-imports-api"
import { getValue } from "src/api/account/kv-api"
import { ProgressUpdate, Timestamp } from "src/interfaces"
import { beforeAll, expect, it, vi } from "vitest"

const accountName = Math.random().toString(36).substring(7)

const mocks = vi.hoisted(() => {
  return {
    access: vi.fn().mockResolvedValue(true),
    readFile: vi.fn(),
  }
})

vi.mock("fs/promises", () => ({
  ...fs.promises,
  access: mocks.access,
  readFile: mocks.readFile,
}))

beforeAll(async () => {
  //
  const fileName = "coinmama.csv"
  const filePath = join("test/files", fileName)
  const buffer = await fs.promises.readFile(filePath, "utf8")
  mocks.readFile.mockResolvedValue(buffer)
  await importFile(accountName, {
    createdBy: "user",
    id: 0,
    metadata: {
      lastModified: 0,
      size: buffer.length,
      type: "text/csv",
    },
    name: fileName,
    scheduledAt: 0,
    status: "completed",
  })
})

it.sequential("should not have balances computed", async () => {
  // act
  const auditLogs = await getAuditLogs(accountName)
  // assert
  expect(auditLogs).toMatchSnapshot()
})

it.sequential("should compute historical balances", async () => {
  // arrange
  const until = Date.UTC(2017, 8, 14, 0, 0, 0, 0) // 14 Sep 2017
  // act
  const updates: ProgressUpdate[] = []
  await computeBalances(accountName, { until }, async (state) => updates.push(state))
  // assert
  const balancesCursor = await getValue<Timestamp>(accountName, "balancesCursor", 0)
  expect(balancesCursor).toBe(until)
  expect(updates.join("\n")).toMatchInlineSnapshot(`
    "0,Computing balances for 2 audit logs
    0,Processing logs 1 to 2
    90,Processed 7 daily balances
    95,Setting networth cursor to Dec 31, 1969
    96,Filling balances to reach today
    100,Saved 14 records to disk"
  `)
})

it.sequential("should re-compute today's balances", async () => {
  // arrange
  const until = Date.UTC(2017, 8, 14, 0, 0, 0, 0) // 14 Sep 2017
  // act
  const updates: ProgressUpdate[] = []
  await computeBalances(
    accountName,
    {
      since: 1504742400000,
      until,
    },
    async (state) => updates.push(state)
  )
  // assert
  const balancesCursor = await getValue<Timestamp>(accountName, "balancesCursor", 0)
  expect(balancesCursor).toBe(until)
  expect(updates.join("\n")).toMatchInlineSnapshot(`
    "0,Refreshing balances starting Sep 07, 2017
    0,Computing balances for 1 audit logs
    0,Processing logs 1 to 1
    90,Processed 1 daily balances
    95,Setting networth cursor to Sep 06, 2017
    96,Filling balances to reach today
    100,Saved 8 records to disk"
  `)
})

it.sequential("should fetch historical balances", async () => {
  // arrange
  // act
  const balances = await getBalances(accountName)
  // assert
  expect(balances.length).toMatchInlineSnapshot(`14`)
  expect(balances).toMatchSnapshot()
})

it.sequential("should have balances computed", async () => {
  // act
  const auditLogs = await getAuditLogs(accountName)
  // assert
  expect(auditLogs).toMatchSnapshot()
})

it.sequential("should fetch latest balances without price data", async () => {
  // arrange
  // act
  const balances = await getBalancesAt(accountName)
  // assert
  expect(balances).toMatchSnapshot()
})

it.sequential("should fetch latest balances with price data", async () => {
  // arrange
  await fetchDailyPrices(accountName, [
    {
      id: "coinmama:BTC",
      priceApiId: "coinbase",
      symbol: "BTC",
    },
  ])
  // act
  const balances = await getBalancesAt(accountName)
  // assert
  expect(balances).toMatchSnapshot()
})

it.sequential("should delete all balances", async () => {
  // arrange
  const initialAuditLogs = await getAuditLogs(accountName)
  // act
  await deleteBalances(accountName)
  // assert
  const balances = await getBalances(accountName)
  expect(balances).toMatchInlineSnapshot(`[]`)
  expect(initialAuditLogs).toMatchInlineSnapshot(`
    [
      {
        "assetId": "coinmama:BTC",
        "balance": "0.072958518",
        "change": "0.036979259",
        "fileImportId": "2702913076",
        "id": "2702913076_3604660326_0",
        "importIndex": 0,
        "operation": "Buy with Credit Card",
        "platform": "coinmama",
        "timestamp": 1504789876000,
        "wallet": "Coinmama Spot",
      },
      {
        "assetId": "coinmama:BTC",
        "balance": "0.035979259",
        "change": "0.035979259",
        "fileImportId": "2702913076",
        "id": "2702913076_1607323056_0",
        "importIndex": 1,
        "operation": "Buy with Credit Card",
        "platform": "coinmama",
        "timestamp": 1504267876000,
        "wallet": "Coinmama Spot",
      },
    ]
  `)
  const auditLogs = await getAuditLogs(accountName)
  expect(auditLogs).toMatchInlineSnapshot(`
    [
      {
        "assetId": "coinmama:BTC",
        "change": "0.036979259",
        "fileImportId": "2702913076",
        "id": "2702913076_3604660326_0",
        "importIndex": 0,
        "operation": "Buy with Credit Card",
        "platform": "coinmama",
        "timestamp": 1504789876000,
        "wallet": "Coinmama Spot",
      },
      {
        "assetId": "coinmama:BTC",
        "change": "0.035979259",
        "fileImportId": "2702913076",
        "id": "2702913076_1607323056_0",
        "importIndex": 1,
        "operation": "Buy with Credit Card",
        "platform": "coinmama",
        "timestamp": 1504267876000,
        "wallet": "Coinmama Spot",
      },
    ]
  `)
})
