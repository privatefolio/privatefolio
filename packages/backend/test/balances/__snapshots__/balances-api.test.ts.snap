// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should fetch historical balances 2`] = `
[
  {
    "timestamp": 1504137600000,
  },
  {
    "coinmama:BTC": "0.035979259",
    "timestamp": 1504224000000,
  },
  {
    "coinmama:BTC": "0.035979259",
    "timestamp": 1504310400000,
  },
  {
    "coinmama:BTC": "0.035979259",
    "timestamp": 1504396800000,
  },
  {
    "coinmama:BTC": "0.035979259",
    "timestamp": 1504483200000,
  },
  {
    "coinmama:BTC": "0.035979259",
    "timestamp": 1504569600000,
  },
  {
    "coinmama:BTC": "0.035979259",
    "timestamp": 1504656000000,
  },
  {
    "coinmama:BTC": "0.072958518",
    "timestamp": 1504742400000,
  },
  {
    "coinmama:BTC": "0.072958518",
    "timestamp": 1504828800000,
  },
  {
    "coinmama:BTC": "0.072958518",
    "timestamp": 1504915200000,
  },
  {
    "coinmama:BTC": "0.072958518",
    "timestamp": 1505001600000,
  },
  {
    "coinmama:BTC": "0.072958518",
    "timestamp": 1505088000000,
  },
  {
    "coinmama:BTC": "0.072958518",
    "timestamp": 1505174400000,
  },
  {
    "coinmama:BTC": "0.072958518",
    "timestamp": 1505260800000,
  },
  {
    "coinmama:BTC": "0.072958518",
    "timestamp": 1505347200000,
  },
]
`;

exports[`should fetch latest balances with price data 1`] = `
[
  {
    "assetId": "coinmama:BTC",
    "balance": "0.072958518",
    "balanceN": 0.072958518,
    "id": "1505347200000_coinmama:BTC",
    "price": {
      "close": 3250.4,
      "high": 3923,
      "low": 3230.05,
      "open": 3855.61,
      "time": 1505347200,
      "value": 3250.4,
      "volume": 36363.92501211,
    },
    "value": 237.1443669072,
  },
]
`;

exports[`should fetch latest balances without price data 1`] = `
[
  {
    "assetId": "coinmama:BTC",
    "balance": "0.072958518",
    "balanceN": 0.072958518,
    "id": "1505347200000_coinmama:BTC",
    "price": undefined,
    "value": undefined,
  },
]
`;

exports[`should have balances computed 1`] = `
[
  {
    "assetId": "coinmama:BTC",
    "balance": "0.072958518",
    "balanceWallet": "0.072958518",
    "change": "0.036979259",
    "fileImportId": "2702913076",
    "id": "2702913076_3604660326_0",
    "importIndex": 0,
    "operation": "Buy with Credit Card",
    "platformId": "ex.coinmama",
    "timestamp": 1504789876000,
    "wallet": "Coinmama Spot",
  },
  {
    "assetId": "coinmama:BTC",
    "balance": "0.035979259",
    "balanceWallet": "0.035979259",
    "change": "0.035979259",
    "fileImportId": "2702913076",
    "id": "2702913076_1607323056_0",
    "importIndex": 1,
    "operation": "Buy with Credit Card",
    "platformId": "ex.coinmama",
    "timestamp": 1504267876000,
    "wallet": "Coinmama Spot",
  },
]
`;

exports[`should not have balances computed 1`] = `
[
  {
    "assetId": "coinmama:BTC",
    "change": "0.036979259",
    "fileImportId": "2702913076",
    "id": "2702913076_3604660326_0",
    "importIndex": 0,
    "operation": "Buy with Credit Card",
    "platformId": "ex.coinmama",
    "timestamp": 1504789876000,
    "wallet": "Coinmama Spot",
  },
  {
    "assetId": "coinmama:BTC",
    "change": "0.035979259",
    "fileImportId": "2702913076",
    "id": "2702913076_1607323056_0",
    "importIndex": 1,
    "operation": "Buy with Credit Card",
    "platformId": "ex.coinmama",
    "timestamp": 1504267876000,
    "wallet": "Coinmama Spot",
  },
]
`;
